books/RayTracingTheNextWeek.html

```diff
@@ -2256,7 +2256,319 @@
+actually hit the quadrilateral.
+
+The plane is divided into coordinate regions like so:
+
+  ![Figure [quad-coords]: Quadrilateral coordinates](../images/fig-2.07-quad-coords.jpg)
+
+
+Thus, to see if a point with planar coordinates $(\alpha,\beta)$ lies inside the quadrilateral, it
+needs to meet the following criteria:
+
+  1. $ 0 \leq \alpha \leq 1 $
+  2. $ 0 \leq \beta \leq 1 $
+
+That's the last piece needed to implement quadrilateral primitives.
+
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ C++
+    #ifndef QUAD_H
+    #define QUAD_H
+
+    #include "rtweekend.h"
+
+    #include "hittable.h"
+
+    #include <cmath>
+
+    class quad : public hittable {
+      public:
+        quad(const point3& _corner, const vec3& _sideA, const vec3& _sideB, shared_ptr<material> m)
+          : corner(_corner), side_A(_sideA), side_B(_sideB), mat(m)
+        {
+            auto n = cross(side_A, side_B);
+            normal = unit_vector(n);
+            D = -dot(normal, corner);
+            w = n / dot(n,n);
+
+            set_bounding_box();
+        }
+
+        virtual void set_bounding_box() {
+            bbox = aabb(corner, corner + side_A + side_B).pad();
+        }
+
+        aabb bounding_box() const override { return bbox; }
+
+        virtual bool hit_ab(double a, double b, hit_record& rec) const {
+            // Given the hit point in plane coordinates, return false if it is outside the
+            // primitive, otherwise set the hit record UV coordinates and return true.
+
+            if ((a < 0) || (1 < a) || (b < 0) || (1 < b))
+                return false;
+
+            rec.u = a;
+            rec.v = b;
+            return true;
+        }
+
+        bool hit(const ray& r, interval ray_t, hit_record& rec) const override {
+            auto denom = dot(normal, r.direction());
+
+            // No hit if the ray is parallel to the plane.
+            if (fabs(denom) < 1e-8)
+                return false;
+
+            // Return false if the hit point parameter t is outside the ray interval.
+            auto t = (-D - dot(normal, r.origin())) / denom;
+            if (!ray_t.contains(t))
+                return false;
+
+            // Determine the hit point lies within the planar shape using its plane coordinates.
+            auto intersection = r.at(t);
+            vec3 planar_hitpt_vector = intersection - corner;
+            auto a = dot(w, cross(planar_hitpt_vector, side_B));
+            auto b = dot(w, cross(side_A, planar_hitpt_vector));
+
+            if (!hit_ab(a, b, rec))
+                return false;
+
+            // Ray hits the 2D shape; set the rest of the hit record and return true.
+            rec.t = t;
+            rec.p = intersection;
+            rec.mat = mat;
+            rec.set_face_normal(r, normal);
+
+            return true;
+        }
+
+      protected:
+        point3 corner;
+        vec3 side_A, side_B;
```

Comment:
decided against using u and v?

Reply:
and then using corner instead of origin

Which feels alright to me. The problem is that linear algebra 100% doesn't care what frame something is in and can be defined against any

Reply:
Fixed (?) in latest rename to Q, u, v.